### PR TITLE
Equality for TlvStream

### PR DESCRIPTION
**Pull request by thomash-acinq**
The order of the elements in a TLV stream is an implementation detail that will disappear with serialization. Equality between TlvStream shouldn't depend on this order.

### CODE REVIEW

**Review by pm47 — COMMENTED** [review]
I think the proper way to do that is to change the collection types to `Set`s:

```scala
case class TlvStream[T <: Tlv](records: Set[T], unknown: Set[GenericTlv] = Nil)
```

And handle the ordering in the serialization, because, as you mention, it's a serialization concern.

Playing with equality of case classes is footgunny.

**Comment by t-bast** [comment]
Why do you want to implement equality on tlv streams in the first place?

**Comment by thomash-acinq** [comment]
> Why do you want to implement equality on tlv streams in the first place?

I need it for bolt12 to check that an invoice matches an invoice request (I extract a `TlvStream[InvoiceRequestTlv]` from the invoice and check that it equals the invoice request).

**Comment by t-bast** [comment]
Ok that makes sense, but I wouldn't go with changing the equality implementation of the case class. Since you can only have a single instance of a given tlv in a stream, it makes more sense to do as @pm47 mentioned and use `Set`s. This may complicate slightly the codec (which needs to read/write and ordered `Seq`) but I don't think it should be too hard?

**Comment by thomash-acinq** [comment]
Now using sets.

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala [hunk 1]
```diff
@@ -497,7 +497,7 @@ object ReplyChannelRange {
             checksums: Option[ReplyChannelRangeTlv.EncodedChecksums]): ReplyChannelRange = {
     timestamps.foreach(ts => require(ts.timestamps.length == shortChannelIds.array.length))
     checksums.foreach(cs => require(cs.checksums.length == shortChannelIds.array.length))
-    new ReplyChannelRange(chainHash, firstBlock, numberOfBlocks, syncComplete, shortChannelIds, TlvStream(timestamps.toList ::: checksums.toList))
+    new ReplyChannelRange(chainHash, firstBlock, numberOfBlocks, syncComplete, shortChannelIds, TlvStream(Set(timestamps, checksums).flatten: Set[fr.acinq.eclair.wire.protocol.ReplyChannelRangeTlv]))
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala:500]: nit:
```suggestion
new ReplyChannelRange(chainHash, firstBlock, numberOfBlocks, syncComplete, shortChannelIds, TlvStream(Set(timestamps, checksums).flatten[ReplyChannelRangeTlv]))
```

File: eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/ChannelRelayerSpec.scala [hunk 2]
```diff
@@ -655,10 +655,10 @@ object ChannelRelayerSpec {
   )
 
   def createBlindedPayload(update: ChannelUpdate, isIntroduction: Boolean): ChannelRelay.Blinded = {
-    val tlvs = TlvStream[OnionPaymentPayloadTlv](Seq(
+    val tlvs = TlvStream[OnionPaymentPayloadTlv](Set(
       Some(OnionPaymentPayloadTlv.EncryptedRecipientData(hex"2a")),
       if (isIntroduction) Some(OnionPaymentPayloadTlv.BlindingPoint(randomKey().publicKey)) else None,
-    ).flatten)
+    ).flatten: Set[OnionPaymentPayloadTlv])
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/ChannelRelayerSpec.scala:661]:
```suggestion
    ).flatten[OnionPaymentPayloadTlv])
```

**Review by t-bast — COMMENTED** [review]
This is definitely cleaner, ACK!

File: eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala (hunk 1 above)
Comment (thomash-acinq) [reply eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LightningMessageTypes.scala]: fixed

File: eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/ChannelRelayerSpec.scala (hunk 2 above)
Comment (thomash-acinq) [reply eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/ChannelRelayerSpec.scala]: fixed

**Review by t-bast — APPROVED** [review]
LGTM :100:

**Pull request merged**